Add set_consistent_hashed_upstream

charliechang opened the pull request:
Hi
I added the support of consistent hash by reuse the directive set_hashed_upstream. I am not sure if there is any alternative for this. (Maybe can be done by Lua?)

agentzh commented:
@charliechang Thanks for sharing.

@doujiang24 Will you have a quick look at this PR please? Thanks!

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +178,35 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+        for (i = 0; i < hashConf->nodeLen; i++) {
+            hashConf->hashNodes[i].name = ul->elts[i];
+            hashConf->hashNodes[i].hash = ngx_hash_key_lc(hashConf->hashNodes[i].name->data,
+                                                          hashConf->hashNodes[i].name->len);
```
One real node should have a number of virtual nodes, usually the number is 160, like nginx chash does: https://github.com/nginx/nginx/blob/master/src/http/modules/ngx_http_upstream_hash_module.c#L299

Otherwise, the consistent hash will have uneven distribution.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +178,35 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+        }
+
+        qsort(hashConf->hashNodes, hashConf->nodeLen, sizeof(ngx_http_set_hashed_upstream_consistent_hash_node),
+              (const void *) ngx_http_set_misc_consistent_hash_compare);
```
When we have a large of nodes, (real node num * 160), this can be bad performance cause this may be called per request.

doujiang24 replied:
My fault, this is only called when parsing the conf. So no performance problem.

doujiang24 commented:
@charliechang Thanks for sharing. Yes, I think it's better to be done in Lua by lua-resty-chash.
https://github.com/agentzh/lua-resty-chash

The way does in this PR can be work too, but it can be bad performance.  May be we can give up this PR? Or may be my wrong? Thoughts?

doujiang24 commented:
@charliechang Sorry, I mean bad performance is not so good, but in some simple case it's good enough.

charliechang commented:
@doujiang24 
Thanks for your suggestion. You are right. This method will result in uneven distribution. I will fix it in my side, Regarding the performance, I found there is a const-time implementation of consistent hash. Maybe it can help the problem?
https://github.com/replay/ngx_http_consistent_hash/blob/master/ngx_http_upstream_consistent_hash_module.c

charliechang commented:
@doujiang24 
Anyway, you can give up this PR or we can keep discussion. It is your call. Thanks for review.

doujiang24 commented:
@charliechang Sorry, I make a stupid mistake here, the performance should not be a problem, please forgive me. Let's continue discuss.
I also found some code style things, Better use the ngx-releng tool from the nginx-devel-utils repos to check your code.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,61 @@
     switch (type) {
-    case ngx_http_set_misc_distribution_modula:
-        return (uint32_t) hash % (uint32_t) ul->nelts;
+        case ngx_http_set_misc_distribution_modula:
```
`case` should left align with `switch`

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,61 @@
-ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
-    ndk_upstream_list_t *ul, ngx_http_set_misc_distribution_t type)
-{
+ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash, ngx_http_set_hashed_upstream_conf *hashConf,
```
This line is exceed 80 chars.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,61 @@
+        case ngx_http_set_misc_distribution_modula:
+            return (uint32_t) hash % (uint32_t) hashConf->ul->nelts;
+        case ngx_http_set_misc_consistent_hash:
+            return ngx_http_set_misc_bsearch(hashConf->hashNodes, hashConf->nodeLen, hash) %
```
We usually do not use uppercase in variable name. `hashNodes` should be `hash_nodes`.

charliechang commented:
@doujiang24 
I added the support of virtual nodes and fixed the problems you mentioned above. Please let me know if there is anything I can make it better!

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
 #ifndef DDEBUG
 #define DDEBUG 0
 #endif
+
```
unnecessary blank line here.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
+        ngx_uint_t hash,
+        ngx_http_set_hashed_upstream_conf *hash_conf,
+        ngx_http_set_misc_distribution_t type) {
+
```
two style thing: 1. we usually try to use lesser lines, `log`, `hash` can follow the function name in the same line.
2. `{` should in a new line.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
-{
+ngx_http_set_misc_apply_distribution(
+        ngx_log_t *log,
+        ngx_uint_t hash,
```
we only need four space indent here.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
     switch (type) {
-    case ngx_http_set_misc_distribution_modula:
-        return (uint32_t) hash % (uint32_t) ul->nelts;
+        case ngx_http_set_misc_distribution_modula:
```
`case` should left align with `switch`.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
+    /* impossible to reach here */
+}
+
+uint32_t
```
two blank line needed before function.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +201,54 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
         return NGX_CONF_ERROR;
     }
 
+    hash_conf = ngx_pcalloc(cf->pool,sizeof(ngx_http_set_hashed_upstream_conf));
```
better to check if hash_conf is `NULL`.

doujiang24 replied:
a space needed after `,`

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
-        return (uint32_t) hash % (uint32_t) ul->nelts;
+        case ngx_http_set_misc_distribution_modula:
+            return (uint32_t) hash % (uint32_t) hash_conf->ul->nelts;
+        case ngx_http_set_misc_consistent_hash:
```
a blank line needed before `case`

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -1,27 +1,71 @@
+        }
+        if (target < nodes[l].hash) {
+            n = l;
+        } else {
```
a blank line needed before this line.

doujiang24 commented:
@charliechang 
Nice, it looks good to me now, but some code style things, hope you won't be impatient about this.
You'd better use `ngx-releng` to check the basic code style, this is the recommended way.
Will you add some test case for this? like `t/hashed-upstream.t` in this repos.
Some more you may need :)
1. https://openresty.gitbooks.io/programming-openresty/content/testing/index.html
2. https://github.com/openresty/test-nginx

charliechang commented:
@doujiang24 
No problem, I will check the coding style and add test cases.

charliechang commented:
@doujiang24 
I think I fixed the coding style issues by checking with ngx_releng. Test case is added as well. I also add some error handling in codes.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -5,18 +5,27 @@
+    ngx_http_set_hashed_upstream_conf *hash_conf,
+    ngx_http_set_misc_distribution_t type)
 {
+
```
unnecessary blank line here.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -5,18 +5,27 @@
+        return hash_conf->hash_nodes[
+                ngx_http_set_misc_bsearch(hash_conf->hash_nodes,
+                                          hash_conf->node_len, hash) %
+                (uint32_t) hash_conf->node_len].index;
```
It's better to calculate a local variable first, like `idx = ngx_http_set_misc_bsearch(hash_conf->hash_nodes,  hash_conf->node_len, hash) % (uint32_t) hash_conf->node_len;`
I personally think it will be more readable.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +34,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+    if (node1->hash < node2->hash) {
+        return -1;
+    }
+    else if (node1->hash > node2->hash) {
```
`else if` should follow the previous `}`, not in an new line.
and need a blank line before the `} else if `.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +34,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+    ngx_http_set_hashed_upstream_consistent_hash_node *nodes, uint32_t node_len,
+    uint32_t target)
+{
+
```
unnecessary blank line here.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +34,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+    uint32_t target)
+{
+
+    uint32_t m = 0, n = node_len, l;
```
better add a blank line after this line.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -78,10 +128,19 @@ ngx_http_set_misc_set_hashed_upstream(ngx_http_request_t *r, ngx_str_t *res,
-    hash = ngx_hash_key_lc(key->data, key->len);
+    if (hash_conf->hash_type == ngx_http_set_misc_consistent_hash) {
+        hash = ngx_crc32_long(key->data, key->len);
+    } else {
```
better add a blank line before this line.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -91,15 +150,38 @@ ngx_http_set_misc_set_hashed_upstream(ngx_http_request_t *r, ngx_str_t *res,
+    ngx_command_t *cmd, void *conf)
+{
+    return ngx_http_set_hashed_upstream_hashtype(
+            cf, cmd, conf, ngx_http_set_misc_distribution_modula);
```
I personally think the below style will be better.
```
    return ngx_http_set_hashed_upstream_hashtype(cf, cmd, conf,
                                               ngx_http_set_misc_distribution_modula);
```

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +228,58 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+    if (ul->nelts <= 0) {
+        hash_conf->node_len = 0;
+        hash_conf->hash_nodes = NULL;
+    } else if (hash_type == ngx_http_set_misc_consistent_hash) {
```
need a blank line before it.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +228,58 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+        hash_conf->node_len = ul->nelts * HASH_VNODES;
+        hash_conf->hash_nodes = ngx_pcalloc(cf->pool,
+            sizeof(ngx_http_set_hashed_upstream_consistent_hash_node) *
+            hash_conf->node_len);
```
we usually calculate a local variable length first.

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +228,58 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+            }
+        }
+
+        qsort(hash_conf->hash_nodes, hash_conf->node_len,
```
better use `ngx_qsort` instead `qsort` for the sake of consistency :)

doujiang24 commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +228,58 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+
+        for (i = 0, j = 1; j < hash_conf->node_len; j++) {
+            if (hash_conf->hash_nodes[i].hash !=
+                hash_conf->hash_nodes[j].hash) {
```
`{` should in a new line

doujiang24 commented on src/ngx_http_set_hashed_upstream.h:
```diff
@@ -10,18 +10,51 @@
-        ngx_command_t *cmd, void *conf);
+
+ngx_uint_t ngx_http_set_misc_apply_distribution
+        (ngx_log_t *log, ngx_uint_t hash,
```
this line may appended to the previous line?

doujiang24 commented:
@charliechang 
Thanks for your patience, I still find some code style thing though the `ngx-releng` can't check it out.
And we may and another test case: one more upstream and the same keys, so we can compare the hash results (just to show the consistency).
Last one thing, will you add some document in the `README.markdown`? thanks :)

charliechang commented:
@doujiang24 
I think I fixed the coding style accordingly. One test case is added for checking consistency. README.markdown is updated as well. Thanks for your review.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +35,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+    ngx_http_set_hashed_upstream_consistent_hash_node *nodes, uint32_t node_len,
+    uint32_t target)
+{
+    uint32_t m = 0, n = node_len, l;
```
Style: nginx coding style requires each C function local variable to be defined in separate lines.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +35,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+
+    while (m < n) {
+        l = (m + n) >> 1;
+        if (target == nodes[l].hash) {
```
Style: need a blank line before block statements like `if`.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +35,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+        if (target == nodes[l].hash) {
+            return l;
+        }
+        if (target < nodes[l].hash) {
```
Ditto. Please fix other similar places.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +35,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+        } else {
+            m = l + 1;
+        }
+    }
```
Style: need a blank line after block statements too.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -25,32 +35,73 @@ ngx_http_set_misc_apply_distribution(ngx_log_t *log, ngx_uint_t hash,
+    if (node1->hash < node2->hash) {
+        return -1;
+
+    } else if (node1->hash > node2->hash) {
```
Because the previous branches always return, we can avoid the `else if` part here.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -91,15 +152,39 @@ ngx_http_set_misc_set_hashed_upstream(ngx_http_request_t *r, ngx_str_t *res,
-    ndk_upstream_list_t     *ul;
-    ngx_str_t               *v;
+    return ngx_http_set_hashed_upstream_hashtype(cf, cmd, conf,
+            ngx_http_set_misc_distribution_modula);
```
Style: argument alignment.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -91,15 +152,39 @@ ngx_http_set_misc_set_hashed_upstream(ngx_http_request_t *r, ngx_str_t *res,
+    void *conf)
+{
+    return ngx_http_set_hashed_upstream_hashtype(cf, cmd, conf,
+            ngx_http_set_misc_consistent_hash);
```
Ditto.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +231,61 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+        hash_conf->hash_nodes = ngx_pcalloc(cf->pool, hash_nodes_size);
+        if (hash_conf->hash_nodes == NULL) {
+            return NGX_CONF_ERROR;
+        }
```
Style: we have indentation problems above.

agentzh commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +231,61 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+
+        for (i = 0, j = 1; j < hash_conf->node_len; j++) {
+            if (hash_conf->hash_nodes[i].hash !=
+                hash_conf->hash_nodes[j].hash)
```
Style: `!=` should be on the second line.

agentzh commented:
@charliechang There are so many coding style issues. Please check the existing code for the right style. Thank you for your contribution!

charliechang commented on src/ngx_http_set_hashed_upstream.c:
```diff
@@ -129,10 +241,67 @@ ngx_http_set_hashed_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+
+        if (hash_conf->hash_nodes == NULL) {
+            return NGX_CONF_ERROR;
+        }
```
@agentzh Could you point out where are the indent problems around here? Thanks

agentzh replied:
@charliechang Sorry, my mistake. I thought the long line breaks were indentation problems.

agentzh commented on src/ngx_http_set_misc_module.c:
```diff
@@ -327,7 +327,16 @@ static ngx_command_t  ngx_http_set_misc_commands[] = {
+    {
+        ngx_string("set_consistent_hashed_upstream"),
+        NGX_HTTP_MAIN_CONF | NGX_HTTP_SRV_CONF | NGX_HTTP_SIF_CONF
+        | NGX_HTTP_LOC_CONF | NGX_HTTP_LIF_CONF | NGX_CONF_TAKE3,
```
Style: indentation inconsistency.

agentzh commented on src/ngx_http_set_misc_module.c:
```diff
@@ -327,7 +327,16 @@ static ngx_command_t  ngx_http_set_misc_commands[] = {
+    },
+    {
+        ngx_string("set_consistent_hashed_upstream"),
+        NGX_HTTP_MAIN_CONF | NGX_HTTP_SRV_CONF | NGX_HTTP_SIF_CONF
```
Style: extra spaces around `|`. Better be consistent with the rest of the code.

agentzh commented on t/consistent-hashed-upstream.t:
```diff
@@ -0,0 +1,90 @@
+use lib 'lib';
+use Test::Nginx::Socket;
+
+#repeat_each(3);
```
Any reason for not enabling `repeat_each`? Is it a leftover of local debugging or a thing done on purpose due to the nature of the tests?

charliechang replied:
Basically the layout of test case is following [hashed-upstream.t]((https://github.com/openresty/set-misc-nginx-module/blob/master/t/hashed-upstream.t)

charliechang commented:
@agentzh I fixed the coding style issues accordingly. Please check it again. Thanks!

charliechang commented:
@agentzh 
Is this PR still valid? since it is been a while...

jbergstroem commented:
I this this would be a good addition. Any comments on why it stalled? Better implemented elsewhere?

agentzh commented:
@jbergstroem Yeah, now that we have `balancer_by_lua*` and such hashing is already implemented in `lua-resty-balancer`:

https://github.com/agentzh/lua-resty-balancer/

It's way more flexible and powerful.

jbergstroem commented:
> @agentzh said:
> @jbergstroem Yeah, now that we have `balancer_by_lua*` and such hashing is already implemented in `lua-resty-balancer`

Of course. Thanks for the pointer. So much to learn! [suggesting we close this then]
